### PR TITLE
2D Image Segmentation of MR images of pelvis using UNet

### PR DESCRIPTION
**Pull request by campbell-mcfadden**
This deep learning model uses the U-Net architecture to segment MR images of the male pelvis. Built using TensorFlow and Keras.

Details instructions on how to train and test the model locally are contained in the README.md file. This file details the model implementation and training process, and gives an evaluation of its performance. It also contains sample image segmentation outputs on the test dataset.

- train.py contains the code required to train the U-Net model
- modules.py contains the blocks of the U-Net model and loss functions
- dataset.py contains the code required to load the NIfTI files
- predict.py contains the code required predict the segmentation of an image in the test dataset
- util.py contains additional helper functions used throughout the code
- paths.py contains the default paths to the data files in the root directory
- the assets folder contains all images used in the readme file

s4644427

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: The problem was solved very well, all labels reached over 0.75 DICE score. All experiments were well recored with DICE and loss plots. Excellent dataset and label distribution analysis.
2. **Implementation**: All functions are well-implemented. It would be better to implement UNet in Class. Hyperparameters are hard-coded rather than using ArgParse.
3. **Design**: Nice
4. **Comment**: insufficient comments for functions. Many functions do not have docstring.
5. **Difficuclty**: **Easy**

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. </td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. </td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Somewhat less.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). Submitted on time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-14</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally
